### PR TITLE
[system-probe] Use ebpf conntracker for prebuilt tracer

### PR DESCRIPTION
**Pull request by hmahmood**
### What does this PR do?

Makes the ebpf based conntracker the default for prebuilt network tracer. We fallback to the netlink based conntracker if loading the ebpf based one fails.

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/tracer/tracer_linux_test.go [hunk 1]
```diff
@@ -1695,3 +1699,101 @@ func TestPreexistingConnectionDirection(t *testing.T) {
...
+func TestEbpfConntrackerFallback(t *testing.T) {
+	var tests = []struct {
+		enableRuntimeCompiler    bool
+		allowPrecompiledFallback bool
+		rcError                  bool
+		prebuiltError            bool
+
+		err        error
+		isPrebuilt bool
+	}{
```
Comment (nplanel) [inline pkg/network/tracer/tracer_linux_test.go:1712]: :+1:

File: pkg/network/tracer/tracer_linux_test.go [hunk 2]
```diff
@@ -1695,3 +1699,101 @@ func TestPreexistingConnectionDirection(t *testing.T) {
...
+			assert.Error(t, err)
+			assert.Nil(t, conntracker)
+			continue
+		}
+
+		assert.NoError(t, err)
+		require.NotNil(t, conntracker)
+		conntracker.Close()
+	}
+}
```
Comment (nplanel) [inline pkg/network/tracer/tracer_linux_test.go:1799]: out of scope of this PR, we should double check everything is ok even if we guess them by dumping the runtime offsets (via a special program (offset-guess-runtime) only to run under our CI) and compare them with prebuilt one
WDYT ?

**Review by nplanel — APPROVED** [review]
LGTM just nit stuff
(out of scope of this PR) May we should double check offsets guess in our CI (I left a comment)

File: pkg/network/tracer/tracer_linux_test.go (hunk 2 above)
Comment (hmahmood) [reply pkg/network/tracer/tracer_linux_test.go:1799]: Done here: https://github.com/DataDog/datadog-agent/pull/16010/files#diff-0f93f50bd50a9b3c0219d97e17c0e8931eaf7a9250b69464348bf031dd5a27ebR143

File: pkg/network/ebpf/c/ip.h [hunk 3]
```diff
@@ -189,11 +189,25 @@ __maybe_unused static __always_inline void flip_tuple(conn_tuple_t *t) {
 // On older kernels, clang can generate Wunused-function warnings on static inline functions defined in
 // header files, even if they are later used in source files. __maybe_unused prevents that issue
 __maybe_unused static __always_inline void print_ip(u64 ip_h, u64 ip_l, u16 port, u32 metadata) {
+// support for %pI4 and %pI6 added in https://github.com/torvalds/linux/commit/d9c9e4db186ab4d81f84e6f22b225d333b9424e3
+#if defined(LINUX_VERSION_CODE) && LINUX_VERSION_CODE >= KERNEL_VERSION(5, 13, 0)
```
Comment (brycekahle) [inline pkg/network/ebpf/c/ip.h:193]: This will be true for prebuilt programs, if the kernel used to compile is 5.13+. It could then run on a kernel < 5.13. You might need to use some of the compile mode defines to protect this.

File: pkg/network/ebpf/offsetguess_types.go [hunk 4]
```diff
@@ -19,15 +19,18 @@ const ProcCommMaxLen = C.TASK_COMM_LEN - 1
 
 type TracerStatus C.tracer_status_t
 
-type TracerState uint8
+type State uint8
```
Comment (brycekahle) [inline pkg/network/ebpf/offsetguess_types.go:22]: `State` feels a bit generic. maybe `OffsetGuessState`?

File: pkg/network/netlink/consumer.go [hunk 5]
```diff
@@ -449,7 +449,6 @@ func (c *Consumer) GetStats() map[string]int64 {
 func (c *Consumer) Stop() {
 	if c.conn != nil {
 		c.conn.Close()
-		c.socket = nil
```
Comment (brycekahle) [inline pkg/network/netlink/consumer.go:452]: Why was this removed?

File: pkg/network/tracer/ebpf_conntracker.go [hunk 6]
```diff
@@ -451,3 +497,39 @@ func getManager(cfg *config.Config, buf io.ReaderAt, maxStateSize int, mapErrTel
 	}
 	return mgr, nil
 }
+
+// getpid returns the current process ID in the host namespace
+func getpid() int32 {
```
Comment (brycekahle) [inline pkg/network/tracer/ebpf_conntracker.go:502]: Can you use `pkg/process/util.GetRootNSPID()` instead?

File: pkg/network/tracer/tracer_linux_test.go [hunk 7]
```diff
@@ -1695,3 +1699,101 @@ func TestPreexistingConnectionDirection(t *testing.T) {
...
+	cfg := testConfig()
+	constants, err := getTracerOffsets(t, cfg)
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		ebpfConntrackerPrebuiltCreator = getPrebuiltConntracker
+		ebpfConntrackerRCCreator = getRuntimeCompiledConntracker
+	})
+
+	for _, te := range tests {
+		t.Logf("%+v", te)
```
Comment (brycekahle) [inline pkg/network/tracer/tracer_linux_test.go:1740]: suggestion: use `t.Run()` instead

File: pkg/network/tracer/tracer_linux_test.go [hunk 8]
```diff
@@ -1695,3 +1699,101 @@ func TestPreexistingConnectionDirection(t *testing.T) {
...
+		conntracker, err := NewEBPFConntracker(cfg, nil, constants)
+		if te.err != nil {
+			assert.Error(t, err)
+			assert.Nil(t, conntracker)
+			continue
+		}
+
+		assert.NoError(t, err)
+		require.NotNil(t, conntracker)
+		assert.Equal(t, te.isPrebuilt, conntracker.(*ebpfConntracker).isPrebuilt)
```
Comment (brycekahle) [inline pkg/network/tracer/tracer_linux_test.go:1764]:
```suggestion
		if assert.NotNil(t, conntracker) {
		     assert.Equal(t, te.isPrebuilt, conntracker.(*ebpfConntracker).isPrebuilt)
		}
```

File: pkg/network/tracer/tracer_linux_test.go [hunk 9]
```diff
@@ -1695,3 +1699,101 @@ func TestPreexistingConnectionDirection(t *testing.T) {
...
+
+		conntracker, err := NewEBPFConntracker(cfg, nil, constants)
+		if te.err != nil {
+			assert.Error(t, err)
+			assert.Nil(t, conntracker)
+			continue
+		}
+
+		assert.NoError(t, err)
+		require.NotNil(t, conntracker)
```
Comment (brycekahle) [inline pkg/network/tracer/tracer_linux_test.go:1763]: This failing will skip the rest of the tests, hence my suggestion to use `t.Run()` above.

File: pkg/network/ebpf/offsetguess_types.go (hunk 4 above)
Comment (brycekahle) [reply pkg/network/ebpf/offsetguess_types.go:22]: Alternatively, could this move into the new `offsetguess` package?

File: pkg/network/tracer/tracer.go [hunk 10]
```diff
@@ -303,55 +286,30 @@ func newReverseDNS(c *config.Config) dns.ReverseDNS {
...
 	}()
-	start := time.Now()
-	editors, err := guessOffsets(offsetMgr, config)
-	if err != nil {
-		return nil, err
+
+	// Offset guessing has been flaky for some customers, so if it fails we'll retry it up to 5 times
+	for i := 0; i < 5; i++ {
+		start := time.Now()
+		if editors, err = guesser.Guess(config); err == nil {
```
Comment (brycekahle) [inline pkg/network/tracer/tracer.go:304]: Will this use stale state from the previous run? Before, we would re-create everything.

**Review by KSerrania — APPROVED** [review]
LGTM for Agent Platform-owned files.

File: pkg/network/netlink/consumer.go (hunk 5 above)
Comment (hmahmood) [reply pkg/network/netlink/consumer.go:452]: This was causing a crash if we closed the consumer before it started receiving events, which happens in some tests.

File: pkg/network/tracer/tracer.go (hunk 10 above)
Comment (hmahmood) [reply pkg/network/tracer/tracer.go:304]: `Guess` does reset the offset guess state

File: pkg/network/tracer/tracer_linux_test.go [hunk 11]
```diff
@@ -1737,32 +1739,35 @@ func TestEbpfConntrackerFallback(t *testing.T) {
 	})
 
 	for _, te := range tests {
-		t.Logf("%+v", te)
-		cfg.EnableRuntimeCompiler = te.enableRuntimeCompiler
-		cfg.AllowPrecompiledFallback = te.allowPrecompiledFallback
+		t.Run("", func(t *testing.T) {
```
Comment (brycekahle) [inline pkg/network/tracer/tracer_linux_test.go:1742]: nit: a short descriptive string would be helpful when interpreting test run data (like CI App Visibility)

**Review by brycekahle — APPROVED** [review]

File: pkg/network/netlink/consumer.go (hunk 5 above)
Comment (brycekahle) [reply pkg/network/netlink/consumer.go:452]: Hmm. Do you have a stack trace? That feels like a problem is something is trying to use the socket after it is supposed to be closed.

Comment (hmahmood) [reply pkg/network/netlink/consumer.go:452]: The consumer read loop in a go routine: https://github.com/DataDog/datadog-agent/blob/main/pkg/network/netlink/consumer.go#L170

File: pkg/network/tracer/tracer_linux_test.go (hunk 11 above)
Comment (hmahmood) [reply pkg/network/tracer/tracer_linux_test.go:1742]: Struggling to come with a short description for the cases :-)
